src/extract_gpuinfo_msm.c

```diff
@@ -0,0 +1,515 @@
+#define STRINGIFY_HELPER_(x) #x
+
+__attribute__((constructor)) static void init_extract_gpuinfo_msm(void) { register_gpu_vendor(&gpu_vendor_msm); }
+
+bool gpuinfo_msm_init(void) {
+  libdrm_handle = dlopen("libdrm.so", RTLD_LAZY);
+  if (!libdrm_handle)
+    libdrm_handle = dlopen("libdrm.so.2", RTLD_LAZY);
+  if (!libdrm_handle)
+    libdrm_handle = dlopen("libdrm.so.1", RTLD_LAZY);
+  if (!libdrm_handle) {
+    local_error_string = dlerror();
+    return false;
+  }
+
+  _drmGetDevices2 = dlsym(libdrm_handle, "drmGetDevices2");
+  if (!_drmGetDevices2)
+    _drmGetDevices = dlsym(libdrm_handle, "drmGetDevices");
+  if (!_drmGetDevices2 && !_drmGetDevices)
+    goto init_error_clean_exit;
+
+  _drmFreeDevices = dlsym(libdrm_handle, "drmFreeDevices");
+  if (!_drmFreeDevices)
+    goto init_error_clean_exit;
+
+  _drmGetVersion = dlsym(libdrm_handle, "drmGetVersion");
+  if (!_drmGetVersion)
+    goto init_error_clean_exit;
+
+  _drmFreeVersion = dlsym(libdrm_handle, "drmFreeVersion");
+  if (!_drmFreeVersion)
+    goto init_error_clean_exit;
+
+  _drmGetMagic = dlsym(libdrm_handle, "drmGetMagic");
+  if (!_drmGetMagic)
+    goto init_error_clean_exit;
+
+  _drmAuthMagic = dlsym(libdrm_handle, "drmAuthMagic");
+  if (!_drmAuthMagic)
+    goto init_error_clean_exit;
+
+  _drmDropMaster = dlsym(libdrm_handle, "drmDropMaster");
+  if (!_drmDropMaster)
+    goto init_error_clean_exit;
+
+  _drmCommandWriteRead = dlsym(libdrm_handle, "drmCommandWriteRead");
+  if (!_drmCommandWriteRead)
+    goto init_error_clean_exit;
+
+  local_error_string = NULL;
+  return true;
+
+init_error_clean_exit:
+  dlclose(libdrm_handle);
+  libdrm_handle = NULL;
+  return false;
+}
+
+void gpuinfo_msm_shutdown(void) {
+  for (unsigned i = 0; i < msm_gpu_count; ++i) {
+    struct gpu_info_msm *current = &gpu_infos[i];
+    _drmFreeVersion(current->drmVersion);
+  }
+
+  free(gpu_infos);
+  gpu_infos = NULL;
+  msm_gpu_count = 0;
+
+  if (libdrm_handle) {
+    dlclose(libdrm_handle);
+    libdrm_handle = NULL;
+    local_error_string = didnt_call_gpuinfo_init;
+  }
+}
+
+static const char *gpuinfo_msm_last_error_string(void) {
+  if (local_error_string) {
+    return local_error_string;
+  } else if (last_libdrm_return_status < 0) {
+    switch (last_libdrm_return_status) {
+    case DRM_ERR_NO_DEVICE:
+      return "no device\n";
+    case DRM_ERR_NO_ACCESS:
+      return "no access\n";
+    case DRM_ERR_NOT_ROOT:
+      return "not root\n";
+    case DRM_ERR_INVALID:
+      return "invalid args\n";
+    case DRM_ERR_NO_FD:
+      return "no fd\n";
+    default:
+      return "unknown error\n";
+    }
+  } else {
+    return "An unanticipated error occurred while accessing AMDGPU "
+           "information\n";
+  }
+}
+
+static const char drm_msm_engine_gpu[] = "drm-engine-gpu";
+static const char drm_msm_cycles_gpu[] = "drm-cycles-gpu";
+static const char drm_msm_maxfreq_gpu[] = "drm-maxfreq-gpu";
+static const char drm_msm_resident_mem[] = "drm-resident-memory";
```

Comment:
fyi, patchset adding drm-resident-memory has not landed _yet_

https://patchwork.freedesktop.org/series/116216/